bsdlib/doc/pdn_extension.rst

```diff
@@ -0,0 +1,174 @@
+.. _pdn_extension:
+
+PDN management
+##############
+
+The BSD library supports configuring up to 10 packet data network (PDN) sockets.
+PDN sockets are an extension to the standard socket API.
+They allow for configuring PDN connections as identified by the user's Access Point Name (APN).
+
+After a PDN socket is configured, it can be used by a data socket (IP or RAW socket).
+The purpose of PDN sockets is to dictate where to send the data, or to do DNS queries on the PDN that is associated with the given APN.
+
+The following sections demonstrate pseudo code for typical use cases of PDN sockets.
+
+Creating a PDN socket
+*********************
+
+.. code-block:: c
+
+   pdn_fd = nrf_socket(NRF_AF_LTE, NRF_SOCK_MGMT, NRF_NPROTO_PDN);
+   nrf_connect(pdn_fd, (struct nrf_sockaddr *)"custom.apn", strlen("custom.apn"));
+
+When creating a PDN socket, the library requests a context ID (CID) to be created.
+It then configures the CID, which creates a PDN.
+This PDN is used to handle data that is sent over the connection.
+
+The example code shows how to create a PDN socket with the protocol family NRF_NPROTO_PDN and the proprietary address family NRF_AF_LTE.
+When the ``nrf_socket`` function returns, a CID has been allocated.
+Next, the ``nrf_connect`` function is used to configure the created CID to use the application's APN.
+If this operation is successful, the library attempts to activate the associated PDN handle.
+If this activation succeeds, ``nrf_connect`` returns with success.
+
+The file descriptor that is returned can only be used to operate on the PDN socket itself.
+For example, you can use it to set and retrieve socket options, to connect, or to close the PDN socket.
+To use the configured PDN, you must use the associated APN name to identify which PDN to use to transfer data (IP/RAW/DNS data).
+See also `Binding a data socket to an APN`_.
+
+Closing a PDN socket
+********************
+
+.. code-block:: c
+
+   nrf_close(pdn_fd);
+
+The close function for PDN sockets de-registers the associated CID/PDN by deactivating the PDN.
+It also removes the configuration of the CID.
+
+Any socket that has been bound to the APN/PDN must be closed before closing the PDN socket.
+Otherwise, it will not have any transport and will start reporting errors.
+
+Binding a data socket to an APN
+*******************************
+
+.. code-block:: c
+
+   tcp_fd = nrf_socket(NRF_AF_INET, NRF_SOCK_STREAM, NRF_IPPROTO_TCP);
+   nrf_setsockopt(tcp_fd, NRF_SOL_SOCKET, NRF_SO_BINDTODEVICE, "custom.apn", strlen("custom.apn"));
+   nrf_bind(...);
```

Comment:
This line is incorrect, we should remove it.

Reply:
OK, removed. :)